fix special FQ node

usstq opened the pull request:
### Details:
 - *FQ node with input_low==input_high was observed in many INT8 models, according to definition, it performs binarization instead of normal quantization, in this PR we trying to support this kind of special FQ node using normal procedure*
 - *...*

### Tickets:
 - *CVS-97942*
 - *CVS-98850*

dmitry-gorokhov commented:
@tomdol  @mateusztabaka  could you please take a look on ONNX FE changes?

tomdol reviewed (approved):
Overall the code is correct but I think it should rather be moved to some optimization transformation like NOP-elimination (or a new dedicated transform which replaces an obsolete FQ node). Please leave it as is and create a ticket to move the code to a transformation or address in this PR.
Also I think at least a single test for this corner case should be added. The ONNX quantized operators are tested here: src/frontends/onnx/tests/onnx_import_quant.in.cpp

zhangYiIntel commented on src/plugins/intel_cpu/src/nodes/fake_quantize.cpp:
```diff
@@ -1896,6 +1896,7 @@ static float roundHalfToEven(float f) {
 }
 
 void FakeQuantize::updateOptimizedFormula(bool do_rounding) {
+    static const float simInfinity = std::pow(2, 100);
```
Just Like the ONNX model, I think the special FQ in IR models also should be supported in LPT transformation which affects less in execution time.

usstq replied:
Thanks a lot, that's good idea, the fix is changed accordingly!

usstq commented:
@tomdol @zhangYiIntel @dmitry-gorokhov Hi all, I changed the solution, according to @tomdol's comment, instead of trying to support such special FQ node at runtime in plugin, I added 2 const folding passes to identify and replace these special FQ nodes with constant+broadcast subgraph which totally eliminates them from the runtime graph.

The root cause of such special FQ node is insufficient calibration dataset used for quantization of models using  POT/nncf tool. for example, when no sample in calibration dataset can produce positive activations in output of Relu/leakyRelu/PRelu layer, the POT/nncf tool will insert FQ with inputLow==inputHigh==outputLow==outputHigh==0, 

```
    f32_[1,1,1,1]  org_1989199316641 = Constant(0)      attrs: element_type=f32 shape=[1,1,1,1] value=?
    f32_[1,1,1,1]  org_1990199415288 = Constant(0)      attrs: element_type=f32 shape=[1,1,1,1] value=?
    f32_[1,1,1,1]  org_1991199516668 = Constant(0)      attrs: element_type=f32 shape=[1,1,1,1] value=?
    f32_[1,1,1,1]  org_1992199616785 = Constant(0)      attrs: element_type=f32 shape=[1,1,1,1] value=?
    f32_[1,64,8,8]  org_Conv_265/WithoutBiases/fq_input_0 = FakeQuantize(org_LeakyRelu_264,org_1989199316641,org_1990199415288,org_1991199516668,org_1992199616785)      attrs: levels=256 auto_broadcast=numpy
```

similarly pytorch/onnx quantization w/o finetuning also procedure quantization/dequantization node with scale equal to zero, and since DequantizeLinear node from onnx will be transformed into multiply(subtract(x, zero_point), scale), this creates pattern such as `y=multiply(x, 0)`

These kind of special FQ node causes trouble in runtime, since FQ with same inputLow & inputHigh requires totally different formula to implement, and more important, inference of such FQ actually is totally waste of time since such computation will finally be Dequantized back to zero,

Thus I introduced two more patterns in LPT's `FoldFakeQuantizeTransformation`:
 1. FQ node with same outputLow/outputHigh inputs: they will be replaced with const-subgraph of outputLow+broadcast;
 2. Multiple of data node (x) with constant all zero tensor (call it `a`): it will be replaced with const-subgraph of a+broadcast;
 
with these changes, 2 models in jira tickets can be inference on plugin w/o any change in plugin, I only tested on CPU, but I guess it should work on other plugins also, I also tested accuracy on 75 quantized models and no regression is found.

dmitry-gorokhov commented:
@v-Golubev  @eshoguli  reassigning to you since now it is LPT related changes.

usstq commented on src/common/low_precision_transformations/src/fold_fake_quantize.cpp:
```diff
@@ -28,35 +32,86 @@ FoldFakeQuantizeTransformation::FoldFakeQuantizeTransformation(const Params& par
-    if (fakeQuantize == nullptr) {
-        return false;
-    }
+    if (fakeQuantize) {
```
test case added!

usstq commented on src/common/low_precision_transformations/src/fold_fake_quantize.cpp:
```diff
@@ -28,35 +32,86 @@ FoldFakeQuantizeTransformation::FoldFakeQuantizeTransformation(const Params& par
-        return false;
+
+    const auto dequantMultiply = ov::as_type_ptr<opset1::Multiply>(m.get_match_root());
+    if (dequantMultiply) {
```
the `dequantizeMultiply`  is actually a normal multiply, not necessarily generated by `FakeQuantizeDecompositionTransformation` , it may comes directly from onnx frontend when translate `DequantizeLinear` op, and it's not FQ, so it only has 2 inputs, right?

usstq commented on src/common/low_precision_transformations/tests/fake_quantize_transformation.cpp:
```diff
@@ -183,13 +183,6 @@ const std::vector<FakeQuantizeTransformationTestValues> fakeQuantizeTransformati
-     ngraph::element::Type_t::i8,
-     {{ngraph::element::f32, {{}, {}, {2.55f}}}},
-     true},
 
```
dot interval FQ is being const-folded in this PR, thus the old compareFunction test case is removed

usstq commented on src/common/low_precision_transformations/tests/fold_fake_quantize_in_transformations.cpp:
```diff
@@ -67,21 +67,34 @@ class FoldFakeQuantizeInTransformations
+                                                                    testValues.constShape,
+                                                                    testValues.actual.constValues);
+        }
 
```
to make sure the test hits the dot-interval const-fold case rather than generic FQ-const-fold, we must use parameter as data input rather than const

usstq commented on src/tests/functional/plugin/gpu/shared_tests_instances/low_precision_transformations/fake_quantize_transformation.cpp:
```diff
@@ -43,10 +43,11 @@ const std::vector<FakeQuantizeTransformationParam> fakeQuantizeOnDataValues = {
+    //    { 256ul, {}, { 0.f }, { 2.55f }, { 2.55f }, { 2.55f } },
+    //    "Pooling", "U8"
+    //},
     {
```
This test cannot success after FQ is being const-folded, because GPU plugin will also const-fold Pooling layer following FQ node, thus Pooling layer is gone.

v-Golubev replied:
Can we leave this test-case but don't check Pooling precision?

usstq replied:
OK

v-Golubev commented on src/common/low_precision_transformations/src/fold_fake_quantize.cpp:
```diff
@@ -28,39 +32,85 @@ FoldFakeQuantizeTransformation::FoldFakeQuantizeTransformation(const Params& par
+    const auto dequantMultiply = ov::as_type_ptr<opset1::Multiply>(m.get_match_root());
+    if (dequantMultiply) {
+        auto multiplyInputs = dequantMultiply->input_values();
+        auto constScales = as_type_ptr<opset1::Constant>(multiplyInputs[0].get_node_shared_ptr());
+        auto dataNode = multiplyInputs[1].get_node_shared_ptr();
+        if (!constScales) {
+            dataNode = multiplyInputs[0].get_node_shared_ptr();
+            constScales = as_type_ptr<opset1::Constant>(multiplyInputs[1].get_node_shared_ptr());
+        }
```
I suggest using `NetworkHelper::getConstantInput` here to get `constScales`

usstq replied:
Sorry, but I also needs dataNode which we cannot deduce easily from `NetworkHelper::getConstantInput`'s return value, thus I prefer to leave it as it is

v-Golubev commented on src/common/low_precision_transformations/src/fold_fake_quantize.cpp:
```diff
@@ -18,7 +19,10 @@ namespace low_precision {
 FoldFakeQuantizeTransformation::FoldFakeQuantizeTransformation(const Params& params) : LayerTransformation(params) {
     MATCHER_SCOPE(FoldFakeQuantizeTransformation);
-    auto fakeQuantize = pattern::wrap_type<opset1::FakeQuantize>();
+    auto matcher = std::make_shared<pattern::op::Or>(OutputVector {
+            pattern::wrap_type<opset1::Multiply>(),
+            pattern::wrap_type<opset1::FakeQuantize>()
+        });
```
We handle in this transformation completely different cases without any common parts: FakeQuantize on constant path or with equal constants, and Multiply with zero. I think this could be a good reason to move Multiply part to separate transformation (something like `FoldZeroMultiply`). @eshoguli what do you think?

usstq replied:
OK, separated

usstq replied:
Looks like we also need someone from `openvino-docs-maintainers` to review

usstq commented on src/common/low_precision_transformations/src/fold_fake_quantize.cpp:
```diff
@@ -28,39 +32,85 @@ FoldFakeQuantizeTransformation::FoldFakeQuantizeTransformation(const Params& par
-    if (ov::is_type<opset1::Constant>(resultConstant)) {
-        replace_node(fakeQuantize, resultConstant);
-        return true;
+    auto fakeQuantizeInputs = fakeQuantize->input_values();
+
+    const auto outputLow = as_type_ptr<opset1::Constant>(fakeQuantizeInputs[3].get_node_shared_ptr());
+    const auto outputHigh = as_type_ptr<opset1::Constant>(fakeQuantizeInputs[4].get_node_shared_ptr());
```
Great!

usstq commented:
@v-Golubev I've fixed the PR according to your comment, can you please review it again ? Thx!

eshoguli reviewed (changes requested):
Offline discussion was started in 98850 ticket. Please, don't merge before the discussion is completed. The general questions:
1. Looks like OV works without changes. 
2. How the model can be working if inputs are not used?
3. Inputs are not removed from the model.
4. Looks like FoldZeroMultiplyTransformation is not used in the ticket model.

v-Golubev reviewed (commented):
Just a few more comments.
Also, due to the fact that these changes might lead to unused model `Parameters`, I suggest validating model caching for the affected models. Thanks!

usstq commented on src/common/low_precision_transformations/src/fold_zero_multiply.cpp:
```diff
@@ -0,0 +1,85 @@
+
+    const auto data_shape_node = fold<opset1::ShapeOf>(dataNode);
+    const auto zero_const =
+        std::make_shared<opset1::Constant>(multiplyNode->get_output_element_type(0), Shape{}, std::vector<size_t>({0}));
```
fold_zero_multiply.cpp was removed according to @eshoguli 's comment

usstq commented on src/common/low_precision_transformations/src/fold_zero_multiply.cpp:
```diff
@@ -0,0 +1,85 @@
+        dataNode = multiplyNode->get_input_node_shared_ptr(1);
+    }
+
+    const auto data_shape_node = fold<opset1::ShapeOf>(dataNode);
```
fold_zero_multiply.cpp was removed according to @eshoguli 's comment

usstq commented:
@eshoguli PR updated, fold_zero_multiply.cpp was removed according to your suggestion, please review again, Thx!

wenjiew commented:
Can we speed up this PR? It is related to a customer model. Thanks!

v-Golubev reviewed (approved):
No comments left from my side.
But I still think that we need to check that models caching behavior isn't affected by these changes (at least, on the models from the tickets)
Also, we need to pass internal performance validation (if it hasn't been passed before) before merge.
Upd. agreed with @eshoguli that functional validation is enough. I made a run that didn't find any functional degradations
